### PR TITLE
Seccomp

### PR DESCRIPTION
**Pull request by xpivarc**
**What this PR does / why we need it**:
This PR aims to solve various use cases of Seccomp with Kubevirt.
There are multiple inputs:
1. Various platforms can enforce or default the Seccomp profile (usually it's some kind of platform default profile) on the admission.
2. Runtime can enforce default Seccomp profile
3. Kubernetes has the option to specify the default Seccompo profile and then use the above points to enforce or default.

Now the problem with the above is:
1. The default or enforced profile can vary drastically. For example, crio denies `userfaultfd`(required to do post-copy) and on the other hand, containerd does not.
2. Various use cases of Kubevirt might want to set the profile on the Kubevirt side arbitrarily or not to set it at all.

This PR solves these problems by allowing to configure Kubevirt profile used for the launcher and at the same time it provides a reference profile.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
It would be great to provide a better experience out of the box. Meaning we should default to our provided profile. As of now, this PR suggests the opposite as it preserves the behavior. I am looking for feedback on this as there is value in breaking the behavior.

**Release note**:
```release-note
Kubevirt can be configured with Seccomp profile. It now ships a custom profile for the launcher.
```

### CODE REVIEW

**Comment by rmohr** [comment]
Curious what you try to acomplish here and what trade-offs you are going to have (like in contrast to selinux policies, having specific container seccomp profiles applied can be a very conscient decision and I wonder if this would e.g. chaing policies, or replace, ...).

**Comment by xpivarc** [comment]
> Curious what you try to acomplish here and what trade-offs you are going to have (like in contrast to selinux policies, having specific container seccomp profiles applied can be a very conscient decision and I wonder if this would e.g. chaing policies, or replace, ...).

Generally, this is aimed to allow Kubevirt to run in an environment that enforces seccomp profile. Now, this can be enforced by various admission webhooks(Pod security admission) or by runtimes themself.  The former can allow custom policies and later forces opinionated profiles(at least from what I saw so far). I used "opinionated" because runtime A can allow "userfaultfd" syscall and runtime B can deny it. Therefore Kubevirt could get into a situation where features or even core functionality might be broken because of the profile.

Now, this to me doesn't look like an ideal situation(patching Kubevirt to allow block features or testing every possible profile...) and as such, I would like to propose maintaining a custom profile for Kubevirt. Initially, I would like to keep it behind the feature gate and see how community needs evolve but for now, I know at least one user of this.

I am not sure what you mean by chaining or replacing. (But if I got you right: )It's possible to append our rules to the runtime profile (at runtime) afaik but only to further block syscalls and not to allow them. Replacing the default/enforced profile seems to me as overstepping Kubernetes.

Does it make sense?

**Comment by alicefr** [comment]
@rmohr do you see a custom seccomp profile differently as having different SELinux policies?
As information, there is a new kubernetes operator ([security-profile-operator](https://github.com/kubernetes-sigs/security-profiles-operator)) that helps defining and managing selinux, seccomp and AppArmor profiles.

**Comment by rmohr** [comment]
> @rmohr do you see a custom seccomp profile differently as having different SELinux policies?
> As information, there is a new kubernetes operator ([security-profile-operator](https://github.com/kubernetes-sigs/security-profiles-operator)) that helps defining and managing selinux, seccomp and AppArmor profiles.

I don't see it necessarily more as an issue like the selinux stuff. But I do not understand yet what we try to solve here exactly.

**Comment by rmohr** [comment]
> The problem is that the default seccomp profile used by cri-o or containerd might block certain syscalls needed by KubeVirt. In this case, `userfaultfd` needed for post-migration

Interesting. I thought we had tests for that ...

I think it is in general fine to do this. One potential difference may be that container_t should be the same on all systems (execept version differences) and vendors typically don't change it, while container runtimes ship different seccomp profiles by default and vendors also modify them sometimes. So maybe not something which we want to have by default (unlike selinux, where it would for sure also be better to not have it, but turnaround times for container_t changes are long, so it may be continue to exist to have a knob on our side to "fix" things).

**Comment by xpivarc** [comment]
/retest
@vladikr @rmohr 
I have adjusted this PR per the discussion.

**Comment by xpivarc** [comment]
Rebased

File: api/openapi-spec/swagger.json
Comment (rmohr) [inline api/openapi-spec/swagger.json:17935]: Can we have a doc string for this?

File: api/openapi-spec/swagger.json
Comment (rmohr) [inline api/openapi-spec/swagger.json:17938]: Also here?

File: tests/operator_test.go
Comment (rmohr) [inline tests/operator_test.go:2949]: s/FContext/Context/

File: tests/operator_test.go
Comment (rmohr) [inline tests/operator_test.go:2996]: Do we require a restart now? You can use a config-change callback to install it without handler restart if that is the case.

File: api/openapi-spec/swagger.json
Comment (rmohr) [reply api/openapi-spec/swagger.json:17935]: Maybe we should call this `installKubeVirtSeccompDefaultProfile`?

File: pkg/virt-controller/services/template.go
Comment (rmohr) [inline pkg/virt-controller/services/template.go:476]: If I set `installProfile` to true, would then the kubevirt profile be picked up? If not, would it make sense to automatically set the custom profile path on the pod if no specific launcher profile is specified?

I am not sure if there is value in separating these.


How about in general restructuring the api a little bit and allow a configuration like this:


Install and use the kubevirt-shipped profile (would also be configured on the pod automatically correctly):

````yaml
seccompConfiguration:
  virtualMachineInstanceProfile:
    KubeVirtProfile: {}
````

Or, as a mutually exclusive option:

````yaml
seccompConfiguration:
  virtualMachineInstanceProfile:
    CustomProfile:
      localhostProfile: /my/path/to/profile
````

**Review by rmohr — COMMENTED** [review]
Did an initial quick pass. A few questions and suggestions.

> It would be great to provide a better experience out of the box. Meaning we should default to our provided profile. As of now, this PR suggests the opposite as it preserves the behavior. I am looking for feedback on this as there is value in breaking the behavior.

I am not sure I agree here. Let's take already the current example: We are adding something to a seccomp profile of a specific set of CRIs (I think drio and moby use the policy you use as base). That has already two issues:

 1. The CRI in use may have completely different opinions on what is safe and unsafe
 2. We already right now use it to make the container potentially **less** secure. If we would make it always more secure I would agree that this would be a reasonable default.
 
 Here also my thought on why it is an issue to potentially run with a less restrictive seccomp profile: With actually your non-root work, which is a great step forward for the project, the pods become more and more accessible by users which can create pods in a given namespace, but are not allowed to create privileged pods. If we weaken the the profile, we potentially allow such users to gain unwanted privileges.
 
 Also here in the `userfaultfd` case I would suggest to check why it is in the profile, and either ask to remove the restriction if there is no real security concern behind it. Do we know why it is not allowed?
 
 Finally we ship a **snapshot** of a default seccomp profile of a speicif cri. If a cri fixes a CVE in that area we are affected as well.
 
 Her is why that are critical differences to our selinux handling:
 
  * the selinux profiles are used on all linux systems. They may have different versions, but they are in principal identical coming from the same source and we don't ship the base profile we extend. It is directly on the node.
  * Since the selinux profiles we care about are developed from a central place, we can at least easier put extensions on top and confer with the selinux people if this is in general a problem and our normal flow there is that we extend the node-local profile mostly with things which will be available in newer policy versions anyway.

**Comment by xpivarc** [comment]
Thanks for the feedback!

> Did an initial quick pass. A few questions and suggestions.
> 
> > It would be great to provide a better experience out of the box. Meaning we should default to our provided profile. As of now, this PR suggests the opposite as it preserves the behavior. I am looking for feedback on this as there is value in breaking the behavior.
> 
> I am not sure I agree here. Let's take already the current example: We are adding something to a seccomp profile of a specific set of CRIs (I think drio and moby use the policy you use as base). That has already two issues:
> 
>     1. The CRI in use may have completely different opinions on what is safe and unsafe
> 
>     2. We already right now use it to make the container potentially **less** secure. If we would make it always more secure I would agree that this would be a reasonable default.

By defaulting to our profile we don't take away the option for the platform or CRI to force their profile. The administrator would need to opt-out of our profile rather than opt-in. 

I think platforms right now don't necessarily enforce profiles(or a majority of them, at least to my extent of knowledge) and certainly, you need to configure this when you are using plain Kubernetes. Therefore defaulting would mean here that we apply at least some profile, not none.

> 
> 
> Here also my thought on why it is an issue to potentially run with a less restrictive seccomp profile: With actually your non-root work, which is a great step forward for the project, the pods become more and more accessible by users which can create pods in a given namespace, but are not allowed to create privileged pods. If we weaken the the profile, we potentially allow such users to gain unwanted privileges.

+1 

> 
> Also here in the `userfaultfd` case I would suggest to check why it is in the profile, and either ask to remove the restriction if there is no real security concern behind it. Do we know why it is not allowed?

I am trying to find a more authoritative answer but for now, I got that there were CVEs around the API and the majority of applications are not expecting to use this syscall. 

For our case, we know we need this and therefore I think it is ok to include it in our profile.

Additionally, we explored different ways around this. You can set up the page faulting also with "userfaultfd" device but I was personally against such implementation. It would require changes to qemu and the device would need to be restricted from usage on the cluster level (which is not by default). It is also unfortunate that it would bypass seccomp altogether.

> 
> Finally we ship a **snapshot** of a default seccomp profile of a speicif cri. If a cri fixes a CVE in that area we are affected as well.

With the custom profile, we need to take over the responsibility of course. I think it is reasonable to accept the burden but of course, the maintainers need to decide in the end.

> 
> Her is why that are critical differences to our selinux handling:
> 
>     * the selinux profiles are used on all linux systems. They may have different versions, but they are in principal identical coming from the same source and we don't ship the base profile we extend. It is directly on the node.

I will object here. Distributions may vary and some 'functions"(pardon my lack of proper terms) can be implemented differently. I personally think we lacked the clarity on the base selinux policy that we required. There was some work but I am not sure what is the state of it. @jean-edouard can comment on it.


> 
>     * Since the selinux profiles we care about are developed from a central place, we can at least easier put extensions on top and confer with the selinux people if this is in general a problem and our normal flow there is that we extend the node-local profile mostly with things which will be available in newer policy versions anyway.

File: api/openapi-spec/swagger.json
Comment (xpivarc) [reply api/openapi-spec/swagger.json:17935]: +1 We definitely need doc string for the new API

File: tests/operator_test.go
Comment (xpivarc) [reply tests/operator_test.go:2996]: I did take this approach as I was not sure how we should react in case the feature would be disabled. Should we remove the profile in such a case? Would the user expect it? And finally, should we take some precautions to determine if the profile is provided by us or by admin(and therefore we should not touch it)

File: pkg/virt-controller/services/template.go
Comment (xpivarc) [reply pkg/virt-controller/services/template.go:476]: I like the suggested API. Reworking...

**Comment by alicefr** [comment]
> > It would be great to provide a better experience out of the box. Meaning we should default to our provided profile. As of now, this PR suggests the opposite as it preserves the behavior. I am looking for feedback on this as there is value in breaking the behavior.
> 
> I am not sure I agree here. Let's take already the current example: We are adding something to a seccomp profile of a specific set of CRIs (I think drio and moby use the policy you use as base). That has already two issues:
> 
> 1. The CRI in use may have completely different opinions on what is safe and unsafe
> 2. We already right now use it to make the container potentially **less** secure. If we would make it always more secure I would agree that this would be a reasonable default.
> 

I think here we need to clarify what is the role of seccomp. Is it seccomp used to provide a set of syscalls declared as "secure"? Or do we want to restrict the set of the syscalls that we expect from the workload?
I doubt it is the first. The container engine needs to provide some good defaults, but we cannot expect that this works well for all kinds of workloads. The workload run in the container is like a black box for the container engine.
Potentially, with a custom profile, we could also decide to reduce the set of the syscalls and remove some of the syscalls allowed in the default one, and make the container more "secure".


> Here also my thought on why it is an issue to potentially run with a less restrictive seccomp profile: With actually your non-root work, which is a great step forward for the project, the pods become more and more accessible by users which can create pods in a given namespace, but are not allowed to create privileged pods. If we weaken the the profile, we potentially allow such users to gain unwanted privileges.
> 
> Also here in the `userfaultfd` case I would suggest to check why it is in the profile, and either ask to remove the restriction if there is no real security concern behind it. Do we know why it is not allowed?
> 
> Finally we ship a **snapshot** of a default seccomp profile of a speicif cri. If a cri fixes a CVE in that area we are affected as well.
> 

I don't think the CVE will be fixed by cri, but I believe this will be more of a kernel bug.
Additionally, a custom kubevirt seccomp profile could help to mitigate future CVEs. Let's suppose we discover a vulnerability in KubeVirt, and we want to block a certain syscall (maybe based on the kernel version of the node). This would be possible only with a custom profile, otherwise, we are tight to the profile shipped by cri-o/containerd.

> Her is why that are critical differences to our selinux handling:
> 
> * the selinux profiles are used on all linux systems. They may have different versions, but they are in principal identical coming from the same source and we don't ship the base profile we extend. It is directly on the node.
> * Since the selinux profiles we care about are developed from a central place, we can at least easier put extensions on top and confer with the selinux people if this is in general a problem and our normal flow there is that we extend the node-local profile mostly with things which will be available in newer policy versions anyway.

File: pkg/virt-handler/seccomp/seccomp_test.go
Comment (alicefr) [inline pkg/virt-handler/seccomp/seccomp_test.go:17]: Can we move the directory creation in a BeforeEach statement? I think we are also missing the directory clean-up.

**Comment by alicefr** [comment]
@xpivarc the PR looks good and I've left a couple of comments and questions.
I still have a couple of questions:
1. Do you have any thoughts about upgrading from a version without the seccomp profile? Will the new virt-launcher pod be created with the profile configured?
2. We might encounter a mismatch if the feature gate wasn't previously enabled. AFAIU, virt-launcher instances aren't able to detect the change in kubevirt cr, and they cannot install the seccomp profile accordingly.

**Comment by xpivarc** [comment]
> @xpivarc the PR looks good and I've left a couple of comments and questions. I still have a couple of questions:
> 
>     1. Do you have any thoughts about upgrading from a version without the seccomp profile? Will the new virt-launcher pod be created with the profile configured?

Yes.

> 
>     2. We might encounter a mismatch if the feature gate wasn't previously enabled. AFAIU, virt-launcher instances aren't able to detect the change in kubevirt cr, and they cannot install the seccomp profile accordingly.

Virt-launcher doesn't do anything with Seccomp so far. Only the handler is responsible for installing the Kubevirt profile. Then runtime is responsible for applying the profile on the whole Pod.

File: pkg/virt-handler/seccomp/seccomp.go
Comment (xpivarc) [reply pkg/virt-handler/seccomp/seccomp.go:45]: Yes, in the future it would be great to just list all allowed syscalls and block others by default. This needs a little bit more work tho and extending the default profile seems good enough as starting point to me.
BTW while cri-o uses this profile it is not per se "cri-o`" profile.

File: pkg/virt-handler/seccomp/seccomp_test.go
Comment (xpivarc) [reply pkg/virt-handler/seccomp/seccomp_test.go]: +1 Done

**Comment by alicefr** [comment]
> 
> > ```
> > 2. We might encounter a mismatch if the feature gate wasn't previously enabled. AFAIU, virt-launcher instances aren't able to detect the change in kubevirt cr, and they cannot install the seccomp profile accordingly.
> > ```
> 
> Virt-launcher doesn't do anything with Seccomp so far. Only the handler is responsible for installing the Kubevirt profile. Then runtime is responsible for applying the profile on the whole Pod.

Yes, sorry, I meant virt-handler not virt-launcher.  If you enable the feature gate later and virt-handler has already been started, the profile won't be installed, right?

**Comment by xpivarc** [comment]
> > > ```
> > > 2. We might encounter a mismatch if the feature gate wasn't previously enabled. AFAIU, virt-launcher instances aren't able to detect the change in kubevirt cr, and they cannot install the seccomp profile accordingly.
> > > ```
> > 
> > 
> > Virt-launcher doesn't do anything with Seccomp so far. Only the handler is responsible for installing the Kubevirt profile. Then runtime is responsible for applying the profile on the whole Pod.
> 
> Yes, sorry, I meant virt-handler not virt-launcher. If you enable the feature gate later and virt-handler has already been started, the profile won't be installed, right?

It will be.

File: cmd/virt-handler/virt-handler.go
Comment (vasiliy-ul) [inline cmd/virt-handler/virt-handler.go:520]: The path should probably be absolute, no? 
```suggestion
	installPath := filepath.Join("/proc/1/root", app.KubeletRoot)
```

Comment (xpivarc) [reply cmd/virt-handler/virt-handler.go:520]: It should

Comment (xpivarc) [reply cmd/virt-handler/virt-handler.go]: done

File: pkg/virt-controller/services/template.go
Comment (acardace) [reply pkg/virt-controller/services/template.go:268]: I agree, let's keep it simple.

File: tests/testsuite/kubevirtresource.go
Comment (acardace) [inline tests/testsuite/kubevirtresource.go:106]: If we use our own profile we no longer need to run migration postcopy e2e tests in a privileged namespace, I think this is the right PR to change them to use the default test namespace.

tests/migration_test.go:641
tests/migration_test.go:2147

Comment (xpivarc) [reply tests/testsuite/kubevirtresource.go:106]: There are more things that can be removed after this PR. I would keep it for follow-up if you don't object.

Comment (acardace) [reply tests/testsuite/kubevirtresource.go:106]: sounds good to me.

**Comment by stu-gott** [comment]
@vladikr @jean-edouard This PR needs an approver. Are one of you available?

**Comment by rmohr** [comment]
/hold

unless it is decided if we want to ship our own seccomp profile and how.

**Comment by stu-gott** [comment]
@rmohr why the hold? Do you have concerns that were not addressed above?

**Comment by rmohr** [comment]
> @rmohr why the hold? Do you have concerns that were not addressed above?

There are multiple things, yes. We would at least need the seccomp profile which we ship imported, so that we know what we ship and can review updates to it. I am also still not sold that we should ship a seccomp profile. Need an additional day to sort out a few things with involved people.

**Comment by rmohr** [comment]
/unhold

Works for me as a temporary solution.

**Comment by stu-gott** [comment]
/override pull-kubevirt-fossa

that test is currently known failing

**Comment by stu-gott** [comment]
@acardace I've rebased this PR to fix the merge conflict. Can you review/LGTM?
@jean-edouard can you please double check changes to virt-handler.go to ensure your original commit is intact?

File: pkg/virt-handler/seccomp/seccomp.go
Comment (jean-edouard) [inline pkg/virt-handler/seccomp/seccomp.go:61]: Looks like DefaultProfile is currently hardcoded in the seccomp module, so we know what to expect from it...
However, it seems like if `userdefaultfd` somehow becomes `ActTrap` or `ActKill`, this function won't catch it.
Same thing if it's in a rule with a single name I guess (where the name is under `.Name` and not `.Names`).
But we know it's not the case, and probably never will be, so it's probably not worth addressing until needed.

File: pkg/virt-handler/seccomp/seccomp.go
Comment (jean-edouard) [inline pkg/virt-handler/seccomp/seccomp.go:18]: Looks like `safepath` would be pretty well suited for this function.
It is not needed for now, as all parts of the path seem to be hardcoded in consts, but if any part of it ever becomes configurable we'll definitely have to switch to using `safepath`.

**Review by jean-edouard — COMMENTED** [review]
/approve
Couple comments in-line, nothing blocking though.

File: pkg/virt-handler/seccomp/seccomp.go
Comment (xpivarc) [reply pkg/virt-handler/seccomp/seccomp.go:61]: Yes, indeed. I think we should be good as functional tests will catch this.

File: pkg/virt-handler/seccomp/seccomp.go
Comment (xpivarc) [reply pkg/virt-handler/seccomp/seccomp.go:18]: I myself was thinking about it. I will follow up in the next PR.

**Comment by xpivarc** [comment]
I have changed the field to feature gate to give us more time to think about if this(the profile definition) should be part of Kubevirt or not.
PTAL @acardace @vladikr @rmohr

**Comment by stu-gott** [comment]
/unhold

@xpivarc has had a chance to address comments.

**Review by acardace — APPROVED** [review]

**Pull request merged**
